packages/flutter/lib/src/services/text_input.dart

```diff
@@ -1420,9 +1492,42 @@ class TextInput {
   ///   topmost [AutofillGroup] is getting disposed.
   static void finishAutofillContext({ bool shouldSave = true }) {
     assert(shouldSave != null);
-    TextInput._instance._channel.invokeMethod<void>(
+    _instance._currentSource.finishAutofillContext(shouldSave: shouldSave);
+  }
+}
+
+class _TextInputSource {
+  static MethodChannel? _channel;
+
+  static void setChannel(MethodChannel newChannel) {
+    _channel = newChannel..setMethodCallHandler(_handleTextInputInvocation);
```

Comment:
should we call clean up on the previous channel?

Reply:
This is only used in tests for setting up a testable environment via `@visibleForTesting TextInput.setChannel()`. This will be moved to the upcoming private default implementation of `TextInputSource` so it won't be visible to the outside.